### PR TITLE
add in full cause to thrown errors

### PR DESCRIPTION
**Pull request by courcelan**
this PR (for issue #93) updates the reporting for a thrown Error.

Errors now report with the following:
- `err.message`: the detail of the error and status code
- `err.cause`: info on what was passed to the API

`err.cause` contains:
- `data`: the detail of the error
- `headers`: any headers
- `status`: response status code
- `statusText`: any response status text
- `config`: the configuration used in the fetch
- `params`: any parameters passed to the reques
- `type`: the content type requested
- `url`: object contain the `URL` schema

Please note that any reference to `auth_token` is removed from the resulting cause.

### CODE REVIEW

**Comment by ViolanteCodes** [comment]
@courcelan can you update the commit message to use `fix:` so the CI/CD deploys?

**Comment by courcelan** [comment]
> @courcelan can you update the commit message to use `fix:` so the CI/CD deploys?

done

File: tests/basic.html [hunk 1]
```diff
@@ -43,11 +43,11 @@
         // add the catch to ensure you do not receive a "uncaught (in promise) warning"
         .catch(
           (err) => {
-            console.error(err)
+            console.error(err.message, err.cause)
           }
         );
 
-      butter.content.cancelRequest()
+      // butter.content.cancelRequest()
```
Comment (marcinpece-apptension) [inline tests/basic.html:50]: Is this line and the one below (87) commented on purpose?

File: examples/hooks.html [hunk 2]
```diff
@@ -58,7 +58,7 @@
           }
         )
         .catch(
-          (error) => console.error("post: ", error)
+          (error) => console.error("post: ", err.message, err.cause)
```
Comment (marcinpece-apptension) [inline examples/hooks.html:61]:
```suggestion
          (err) => console.error("post: ", err.message, err.cause)
```

File: examples/hooks.html [hunk 3]
```diff
@@ -67,7 +67,7 @@
           (res) => console.log(res)
         )
         .catch(
-          (error) => console.error("page: ", error)
+          (error) => console.error("page: ", err.message, err.cause)
```
Comment (marcinpece-apptension) [inline examples/hooks.html:70]:
```suggestion
          (err) => console.error("page: ", err.message, err.cause)
```

File: tests/hooks.html [hunk 4]
```diff
@@ -58,7 +58,7 @@
           }
         )
         .catch(
-          (error) => console.error("post: ", error)
+          (error) => console.error("post: ", err.message, err.cause)
```
Comment (marcinpece-apptension) [inline tests/hooks.html:61]:
```suggestion
          (err) => console.error("post: ", err.message, err.cause)
```

File: tests/hooks.html [hunk 5]
```diff
@@ -67,7 +67,7 @@
           (res) => console.log(res)
         )
         .catch(
-          (error) => console.error("page: ", error)
+          (error) => console.error("page: ", err.message, err.cause)
```
Comment (marcinpece-apptension) [inline tests/hooks.html:70]:
```suggestion
          (err) => console.error("page: ", err.message, err.cause)
```

File: lib/useButter.js [hunk 6]
```diff
@@ -123,18 +123,38 @@ export default function useButter(type, butterConfig) {
                     detail: errorDetail
```
Comment (marcinpece-apptension) [inline lib/useButter.js:123]: In some cases error detail might be not in `detail` prop. Should we cover somehow or current state is ok?
https://share.zight.com/8LuNRWmL

**Review by marcinpece-apptension — COMMENTED** [review]
https://share.zight.com/nOu7A25x

File: tests/basic.html (hunk 1 above)
Comment (courcelan) [reply tests/basic.html:50]: nope. just commented out so I could test.

File: examples/hooks.html (hunk 2 above)
Comment (courcelan) [reply examples/hooks.html:61]: thanks for the catch.

File: examples/hooks.html (hunk 3 above)
Comment (courcelan) [reply examples/hooks.html:70]: thanks for the catch.

File: tests/hooks.html (hunk 4 above)
Comment (courcelan) [reply tests/hooks.html:61]: thanks for the catch!

File: tests/hooks.html (hunk 5 above)
Comment (courcelan) [reply tests/hooks.html:70]: thanks for the catch!

File: lib/useButter.js (hunk 6 above)
Comment (courcelan) [reply lib/useButter.js]: thanks for this @marcinpece-apptension, I've updated the code to handle a more general set of error keys/values. Please give this all another check.

File: lib/useButter.js [hunk 7]
```diff
@@ -119,22 +119,57 @@ export default function useButter(type, butterConfig) {
...
+                const cause = {
+                    data: errors,
+                    headers: errorPayload.response.headers,
+                    status: errorPayload.response.status,
+                    statusText: errorPayload.response.statusText,
+                    config: errorPayload.config,
+                    params: mappedParams,
+                    type,
+                    url
+                };
```
Comment (marcinpece-apptension) [inline lib/useButter.js:143]: 💚

File: __tests__/butter.test.js [hunk 8]
```diff
@@ -129,7 +129,7 @@ test(
     catch (error) {
       await expect(error).toBeDefined();
       await expect(error).toBeInstanceOf(Error);
-      await expect(error.message).toEqual("Error: Not found");
+      await expect(error.message).toEqual("Error: Not found (404)");
```
Comment (marcinpece-apptension) [inline __tests__/butter.test.js:132]: I have one error while running tests https://share.zight.com/L1uEo6mm

**Comment by marcinpece-apptension** [comment]
@ViolanteCodes Has [the error](https://github.com/ButterCMS/buttercms-js/pull/94#discussion_r1713354910) in the test been fixed before merge?

**Comment by courcelan** [comment]
@marcinpece-apptension yes. I caught it in #95 and updated there prior to you mentioning it. so we should be all good.

**Comment by marcinpece-apptension** [comment]
Cool 🎉 Thanks

**Pull request merged**